// vue-styleguidist/lib/scripts/create-server.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = createServer;

var _webpack = _interopRequireDefault(require("webpack"));

var _webpackDevServer = _interopRequireDefault(require("webpack-dev-server"));

var _webpackMerge = _interopRequireDefault(require("webpack-merge"));

var _makeWebpackConfig = _interopRequireDefault(require("./make-webpack-config"));

function createServer(config, env) {
  var webpackConfig = (0, _makeWebpackConfig["default"])(config, env);

  var _merge = (0, _webpackMerge["default"])({
    devServer: {
      noInfo: true,
      compress: true,
      clientLogLevel: 'none',
      hot: true,
      quiet: true,
      disableHostCheck: true,
      injectClient: false,
      watchOptions: {
        ignored: /node_modules/
      },
      watchContentBase: config.assetsDir !== undefined,
      stats: webpackConfig.stats || {}
    }
  }, {
    devServer: webpackConfig.devServer
  }, {
    devServer: {
      contentBase: config.assetsDir
    }
  }),
      webpackDevServerConfig = _merge.devServer;

  var compiler = (0, _webpack["default"])(webpackConfig);
  var devServer = new _webpackDevServer["default"](compiler, webpackDevServerConfig); // User defined customizations

  if (config.configureServer) {
    config.configureServer(devServer, env);
  }

  return {
    app: devServer,
    compiler: compiler
  };
}
