// vue-styleguidist/lib/client/rsg-components/SectionHeading/SectionHeading.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import PropTypes from 'prop-types';
import Slot from 'rsg-components/Slot';
import SectionHeadingRenderer from 'rsg-components/SectionHeading/SectionHeadingRenderer';
import getUrl from 'react-styleguidist/lib/client/utils/getUrl';

var SectionHeading = function SectionHeading(_ref) {
  var slotName = _ref.slotName,
      slotProps = _ref.slotProps,
      children = _ref.children,
      id = _ref.id,
      pagePerSection = _ref.pagePerSection,
      rest = _objectWithoutProperties(_ref, ["slotName", "slotProps", "children", "id", "pagePerSection"]);

  var href = pagePerSection ? getUrl({
    slug: id,
    id: rest.depth !== 1,
    takeHash: true
  }) : getUrl({
    slug: id,
    anchor: true
  });
  var parentHref = slotProps.parentComponent && slotProps.parentComponent.href;
  return React.createElement(SectionHeadingRenderer, _extends({
    toolbar: slotProps.parentComponent ? null : React.createElement(Slot, {
      name: slotName,
      props: slotProps
    }),
    id: id,
    href: href
  }, rest, {
    parentName: slotProps.parentComponent && slotProps.parentComponent.visibleName,
    parentHref: parentHref
  }), children);
};

SectionHeading.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string.isRequired,
  slotName: PropTypes.string.isRequired,
  slotProps: PropTypes.shape({
    isolated: PropTypes.bool.isRequired,
    parentComponent: PropTypes.shape({
      href: PropTypes.string,
      visibleName: PropTypes.string.isRequired
    })
  }).isRequired,
  depth: PropTypes.number.isRequired,
  deprecated: PropTypes.bool,
  pagePerSection: PropTypes.bool
};
export default SectionHeading;
