// vue-styleguidist/lib/client/rsg-components/JsDoc/JsDoc.js
import "core-js/modules/es.symbol";
import "core-js/modules/es.symbol.description";
import "core-js/modules/es.array.join";
import "core-js/modules/es.array.map";
import "core-js/modules/es.function.name";
import "core-js/modules/es.object.to-string";
import "core-js/modules/es.regexp.to-string";
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import PropTypes from 'prop-types';
import map from 'lodash/map';
import capitalize from 'lodash/capitalize';
import Markdown from 'rsg-components/Markdown';
import Argument from 'rsg-components/Argument';
import Styled from 'rsg-components/Styled';

var styles = function styles(_ref) {
  var space = _ref.space,
      color = _ref.color;
  return {
    wrapper: {
      color: color.base,
      fontSize: 'inherit',
      lineHeight: 1.5
    },
    name: {
      marginRight: space[1]
    }
  };
};

var list = function list(array) {
  return array.map(function (item) {
    return item.description;
  }).join(', ');
};

var paragraphs = function paragraphs(array) {
  return array.map(function (item) {
    return item.description;
  }).join('\n\n');
};

var fields = {
  deprecated: function deprecated(value) {
    return "".concat(value[0].description);
  },
  see: function see(value) {
    return paragraphs(value);
  },
  link: function link(value) {
    return paragraphs(value);
  },
  author: function author(value) {
    return list(value);
  },
  version: function version(value) {
    return "".concat(value[0].description);
  },
  since: function since(value) {
    return "".concat(value[0].description);
  }
};

var JsDocRenderer = function JsDocRenderer(_ref2) {
  var classes = _ref2.classes,
      field = _ref2.field,
      children = _ref2.children;
  return React.createElement("div", {
    className: "vsg-jsdoc-tag ".concat(classes.wrapper),
    key: field
  }, React.createElement("span", {
    className: "vsg-tag-name ".concat(classes.name)
  }, capitalize(field), ":"), React.createElement("span", {
    className: "vsg-tag-value ".concat(classes.value)
  }, children));
};

export var JsDoc = function JsDoc(_ref3) {
  var classes = _ref3.classes,
      props = _objectWithoutProperties(_ref3, ["classes"]);

  return React.createElement(React.Fragment, null, props["throws"] && props["throws"].map(function (_throws, i) {
    return React.createElement(JsDocRenderer, {
      key: i,
      field: "throws",
      classes: classes
    }, React.createElement(Argument, _extends({
      name: ""
    }, _throws, {
      description: typeof _throws.description === 'boolean' ? _throws.description.toString() : _throws.description
    })));
  }), map(fields, function (format, field) {
    var value = props[field];
    if (!value || !Array.isArray(value)) return null;
    return React.createElement(JsDocRenderer, {
      key: field,
      field: field,
      classes: classes
    }, React.createElement(Markdown, {
      text: format(value) || '',
      inline: true
    }));
  }));
};
JsDoc.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string.isRequired).isRequired,
  deprecated: PropTypes.array,
  see: PropTypes.array,
  link: PropTypes.array,
  author: PropTypes.array,
  version: PropTypes.array,
  since: PropTypes.array,
  "throws": PropTypes.array
};
export default Styled(styles)(JsDoc);
