// vue-styleguidist/lib/client/rsg-components/PlaygroundAsync/PlaygroundAsync.js
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import { Component } from 'react';
import { polyfill } from 'react-lifecycles-compat';
import PropTypes from 'prop-types';
import Context from 'rsg-components/Context';
import Playground from 'react-styleguidist/lib/client/rsg-components/Playground/Playground';

var PlaygroundAsync =
/*#__PURE__*/
function (_Component) {
  _inherits(PlaygroundAsync, _Component);

  function PlaygroundAsync(props, context) {
    var _this;

    _classCallCheck(this, PlaygroundAsync);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(PlaygroundAsync).call(this, props, context));

    _defineProperty(_assertThisInitialized(_this), "componentWillUnmount", Playground.prototype.componentWillUnmount);

    _defineProperty(_assertThisInitialized(_this), "handleChange", Playground.prototype.handleChange);

    _defineProperty(_assertThisInitialized(_this), "handleTabChange", Playground.prototype.handleTabChange);

    _defineProperty(_assertThisInitialized(_this), "getInitialActiveTab", Playground.prototype.getInitialActiveTab);

    Playground.call(_assertThisInitialized(_this), props, context);
    return _this;
  }

  _createClass(PlaygroundAsync, [{
    key: "render",
    value: function render() {
      return Playground.prototype.render.call(this);
    }
  }]);

  return PlaygroundAsync;
}(Component);

_defineProperty(PlaygroundAsync, "propTypes", _objectSpread({}, Playground.propTypes, {
  code: PropTypes.shape({
    raw: PropTypes.string.isRequired,
    compiled: PropTypes.oneOfType([PropTypes.shape({
      script: PropTypes.string,
      template: PropTypes.string,
      style: PropTypes.string
    }), PropTypes.bool])
  }).isRequired
}));

_defineProperty(PlaygroundAsync, "contextType", Context);

_defineProperty(PlaygroundAsync, "getDerivedStateFromProps", Playground.getDerivedStateFromProps);

export default polyfill(PlaygroundAsync);
