// vue-styleguidist/lib/scripts/utils/mergeWebpackConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = mergeWebpackConfig;

var isFunction = require('lodash/isFunction');

var omit = require('lodash/omit');

var mergeBase = require('webpack-merge');

var IGNORE_SECTIONS = ['entry', 'output', 'watch', 'stats', 'styleguidist'];
var IGNORE_SECTIONS_ENV = {
  development: [],
  // For production builds, we'll ignore devtool settings to avoid
  // source mapping bloat.
  production: ['devtool'],
  removePlugins: ['plugins']
};
var IGNORE_PLUGINS = ['CommonsChunkPlugins', 'MiniHtmlWebpackPlugin', 'HtmlWebpackPlugin', 'OccurrenceOrderPlugin', 'DedupePlugin', 'UglifyJsPlugin', 'HotModuleReplacementPlugin'];
var merge = mergeBase({
  // Ignore user’s plugins to avoid duplicates and issues with our plugins
  customizeArray: mergeBase.unique('plugins', IGNORE_PLUGINS, function (plugin) {
    return plugin.constructor && plugin.constructor.name;
  })
}); //make it a typeguard

function isFunc(conf) {
  return isFunction(conf);
}
/**
 * Merge two Webpack configs.
 *
 * In the user config:
 * - Ignores given sections (options.ignore).
 * - Ignores plugins that shouldn’t be used twice or may cause issues.
 *
 * @param {object} baseConfig
 * @param {object|Function} userConfig
 * @param {string} env
 * @return {object}
 */


function mergeWebpackConfig(baseConfig, userConfig, env) {
  var userConfigObject = isFunc(userConfig) ? userConfig(env) : userConfig;
  var safeUserConfig = omit(userConfigObject, IGNORE_SECTIONS.concat(IGNORE_SECTIONS_ENV[env]));
  return merge(baseConfig, safeUserConfig);
}
