// vue-styleguidist/lib/loaders/utils/getParser.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = getParser;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _vueDocgenApi = require("vue-docgen-api");

var _mergeWebpackConfig = _interopRequireDefault(require("../../scripts/utils/mergeWebpackConfig"));

function getParser(config) {
  // resolve webpack config as functions or objects
  var webpackConfig = (0, _mergeWebpackConfig["default"])({}, config.webpackConfig, process.env.NODE_ENV || 'production');
  var alias;
  var modules;

  if (webpackConfig.resolve) {
    alias = webpackConfig.resolve.alias;
    modules = webpackConfig.resolve.modules;
  }

  var defaultParser =
  /*#__PURE__*/
  function () {
    var _ref = (0, _asyncToGenerator2["default"])(
    /*#__PURE__*/
    _regenerator["default"].mark(function _callee(file) {
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return (0, _vueDocgenApi.parse)(file, {
                alias: alias,
                modules: modules,
                jsx: config.jsxInComponents,
                validExtends: config.validExtends
              });

            case 2:
              return _context.abrupt("return", _context.sent);

            case 3:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));

    return function defaultParser(_x) {
      return _ref.apply(this, arguments);
    };
  }();

  return config.propsParser || defaultParser;
}
