// vue-styleguidist/lib/client/rsg-components/EditorWithToolbar/EditorWithToolbar.js
/* eslint-disable react/prop-types */
import React from 'react';
import copy from 'clipboard-copy';
import { MdContentCopy } from 'react-icons/md'; // eslint-disable-next-line import/no-unresolved,import/extensions

import Editor from 'rsg-components/EditorNoTools';
import ToolbarButton from 'rsg-components/ToolbarButton';
import Styled from 'rsg-components/Styled';

var styles = function styles(_ref) {
  var space = _ref.space;
  return {
    container: {
      position: 'relative'
    },
    copyButton: {
      position: 'absolute',
      right: space[1],
      top: space[1],
      zIndex: 3,
      cursor: 'pointer'
    }
  };
};

function EditorWithToolbar(props) {
  var classes = props.classes;
  return React.createElement("div", {
    className: classes.container
  }, React.createElement(ToolbarButton, {
    small: true,
    className: classes.copyButton,
    onClick: function onClick() {
      return copy(props.code);
    },
    title: "Copy to clipboard"
  }, React.createElement(MdContentCopy, null)), React.createElement(Editor, props));
}

export default Styled(styles)(EditorWithToolbar);
